abnf/odata-aggregation-abnf.txt

```diff
@@ -130,15 +130,25 @@ pathPrefix      = [ ( qualifiedEntityTypeName / qualifiedComplexTypeName ) "/" ]
 pathSegment     = ( complexProperty / complexColProperty ) [ "/" qualifiedComplexTypeName ]
                 / navigationProperty [ "/" qualifiedEntityTypeName  ]

```

Comment:
Must we extend `pathSegment` in analogy with `grpPropPathSegment`? Are annotations and type casts also allowed in `aggregate(... from ...)` [OData-Aggr, section 3.1.4]?

Reply:
Several points:
1. The rules for path segments used in groupby and in aggregate are different. It would improve readability to prefix `pathSegment` and `pathPrefix` with `aggr` to clearly indicate their limited scope
2. Re: "Must we extend pathSegment in analogy with grpPropPathSegment?" I see two deviations between the spec and the `pathPrefix`/`pathSegment` rules:
  a) The spec states in 3.1.4: "P is a path made up of [...] single- or collection-valued instance annotations of entity or complex type". I dont see this reflected in the ABNF rules. Isn't `grpPropAnnoSegment ` missing in `pathSegment`? And if so, we should remove the `grp` prefix to indicate its general scope
  b) `pathPrefix` includes type casts, which are not mentioned in the spec. They can be useful, so we should think about adding this to the spec in section 3.1!?
2. Re "Are annotations also allowed in aggregate(... from ...) [OData-Aggr, section 3.1.4]" The `from` clause is defined by reduction to a sequence of groupby/aggregate, where the property in the `from` clause is a grouping property of the input set. In this sense, the complete syntax of `groupingProperty` should be available, incl. annotations. Do you see any problems?
3. Re: "What does a final type cast segment mean?" Good catch, doesn't look correct as type casts in path expressions are always a prefix of the path! Should be removed

Reply:
>    a) The spec states in 3.1.4: "P is a path made up of [...] single- or collection-valued instance annotations of entity or complex type". I dont see this reflected in the ABNF rules. Isn't `grpPropAnnoSegment ` missing in `pathSegment`? And if so, we should remove the `grp` prefix to indicate its general scope

`grpPropAnnoSegment` allows only single-valued annotations, therefore I would not reuse it here. Instead, we need a new rule `aggrAnnoSegment` or similar.

>    b) `pathPrefix` includes type casts, which are not mentioned in the spec. They can be useful, so we should think about adding this to the spec in section 3.1!?

Yes.

> 3. Re "Are annotations also allowed in aggregate(... from ...) [OData-Aggr, section 3.1.4]" The `from` clause is defined by reduction to a sequence of groupby/aggregate, where the property in the `from` clause is a grouping property of the input set. In this sense, the complete syntax of `groupingProperty` should be available, incl. annotations. Do you see any problems?

Cannot judge this so easily.

> 4. Re: "What does a final type cast segment mean?" Good catch, doesn't look correct as type casts in path expressions are always a prefix of the path! Should be removed

OK, can you create a separate Jira issue for all this?

Reply:
https://github.com/oasis-tcs/odata-abnf/pull/75